modyn/models/gpt2/gpt2.py

```diff
@@ -0,0 +1,88 @@
+from typing import Any
+
+import torch
+from torch import nn
+from transformers import AutoTokenizer, GPT2Config, GPT2LMHeadModel
+
+from modyn.models.coreset_methods_support import CoresetSupportingModule
+
+
+class Gpt2:
+    def __init__(self, hparams: Any, device: str, amp: bool) -> None:
+        self.model = Gpt2Modyn(hparams)
+        self.model.to(device)
+
+
+class Gpt2Modyn(CoresetSupportingModule):
+    def __init__(self, hparams: Any) -> None:
+        super().__init__()
+
+        # Set default values for hyperparameters if not provided in hparams
+        model_name = getattr(hparams, "model_name_or_path", "gpt2-large")
+        self.model_name = model_name if model_name in {"gpt2", "gpt2-medium", "gpt2-large", "gpt2-xl"} else "gpt2-large"
+        self.max_length = getattr(hparams, "max_length", 50)
+        self.temperature = getattr(hparams, "temperature", 1.0)
+        self.top_k = getattr(hparams, "top_k", 50)
+        self.top_p = getattr(hparams, "top_p", 0.95)
+        self.num_return_sequences = getattr(hparams, "num_return_sequences", 1)
+        self.enable_flash_attention = getattr(hparams, "enable_flash_attention", True)
+        self.enable_gradient_checkpointing = getattr(hparams, "enable_gradient_checkpointing", False)
+        self.dtype = getattr(hparams, "dtype", torch.float32)
+
+        # Load tokenizer from gpt2-large (or adjust as needed)
+        self.tokenizer = AutoTokenizer.from_pretrained(self.model_name)
+        self.tokenizer.pad_token = self.tokenizer.eos_token  # Use right padding
+
+        # Load the GPT-2 model with FlashAttention if enabled
+        self.model = GPT2LMHeadModel.from_pretrained(self.model_name, torch_dtype=self.dtype)
+
+        if self.enable_flash_attention:
+            self.model.config.attn_implementation = "flash_attention"
+
+        # If gradient checkpointing is enabled, enable it
+        if self.enable_gradient_checkpointing:
+            self.model.gradient_checkpointing_enable()
+
+        # Load model configuration
+        self.config = GPT2Config.from_pretrained(self.model_name)
```

Comment:
What is this object being used for? I am a bit confused since we already have instantiated a model